lib/components/sections/thumbnail_grid/item_section_thumbnail_grid.dart

```diff
@@ -0,0 +1,136 @@
+
+class ItemSectionThumbnailGrid extends StatelessWidget {
+  final Fragment$GridSection data;
+  final double aspectRatio;
+  final bool showSecondaryTitle;
+
+  const ItemSectionThumbnailGrid({
+    required this.data,
+    required this.aspectRatio,
+    required this.showSecondaryTitle,
+    super.key,
+  });
+
+  factory ItemSectionThumbnailGrid.fromDefaultGridSection(Fragment$Section$$DefaultGridSection data) {
+    return ItemSectionThumbnailGrid(
+      data: data,
+      showSecondaryTitle: data.metadata?.secondaryTitles ?? true,
+      aspectRatio: 16 / 9,
+    );
+  }
+
+  factory ItemSectionThumbnailGrid.fromPosterGridSection(Fragment$Section$$PosterGridSection data) {
+    return ItemSectionThumbnailGrid(
+      data: data,
+      showSecondaryTitle: data.metadata?.secondaryTitles ?? true,
+      aspectRatio: 0.67,
+    );
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Container(
+      margin: const EdgeInsets.symmetric(horizontal: 16),
+      child: GridSectionList(
+        gridSize: data.gridSize,
+        sectionItems: data.items.items,
+        showSecondaryTitle: showSecondaryTitle,
+      ),
+    );
+  }
+}
+
+class GridSectionList extends ConsumerWidget {
+  const GridSectionList({
+    super.key,
+    required this.gridSize,
+    this.aspectRatio = 16 / 9,
+    required this.sectionItems,
+    this.showSecondaryTitle = true,
+  });
+
+  final Enum$GridSectionSize gridSize;
+  final List<Fragment$GridSectionItem> sectionItems;
+  final bool showSecondaryTitle;
+  final double aspectRatio;
+
+  int getColumnCount() {
+    switch (gridSize) {
+      case Enum$GridSectionSize.half:
+      default:
+        return 2;
+    }
+  }
+
+  Widget getItemWidget(Fragment$GridSectionItem sectionItem, Fragment$Episode? curLiveEpisode) {
+    var episode = sectionItem.item.asOrNull<Fragment$GridSectionItem$item$$Episode>();
+    if (episode != null) {
+      return ThumbnailGridEpisode(
+        sectionItem: sectionItem,
+        episode: episode,
+        showSecondaryTitle: showSecondaryTitle,
+        aspectRatio: aspectRatio,
+        isLive: sectionItem.id == curLiveEpisode?.id,
+      );
+    }
+    var show = sectionItem.item.asOrNull<Fragment$GridSectionItem$item$$Show>();
+    if (show != null) {
+      return ThumbnailGridShow(
+        sectionItem: sectionItem,
+        show: show,
+        aspectRatio: aspectRatio,
+      );
+    }
+    return const SizedBox();
+  }
+
+  @override
+  Widget build(BuildContext context, WidgetRef ref) {
+    final columnCount = getColumnCount();
+    Fragment$Episode? curLiveEpisode = ref.watch(currentLiveEpisodeProvider)?.episode;
+
+    final items = sectionItems
+        .where((element) => element.item is Fragment$GridSectionItem$item$$Episode || element.item is Fragment$GridSectionItem$item$$Show)
+        .toList();
+    final rowSize = (items.length / columnCount).ceil();
+
+    final rows = List<GridRow>.generate(rowSize, (rowIndex) {
```

Comment:
moving this function into the `itemBuilder` below is a better practice.
The difference is that this runs on **every** build() now, while `itemBuilder(index)` only runs for the subset of `index`es that are on screen.